### PR TITLE
Centralized capabilities service

### LINKED ISSUE #9695: Write to container: permissions vs logic
I have a feeling we need an additional hook to check container write logic. Currently, we use `canWriteToContainer` to check if one entity type can contain another entity type (only discussion can contain discussion replies), which relates more to logic than permissions. The problem with the current approach is that we end up allowing bad logic with ignored access (i.e. admins can end up writing a discussion reply to something other than a discussion). 
This would also help with cases like closed discussions and blog comments being off. Permission wise a user can write discussion replies to a closed discussion, logic wise they shouldn't be allowed to.

### PR DESCRIPTION
**Pull request by hypeJunction**
**feature(entities): adds user capabilities service and tests**
Adds `\Elgg\UserCapabilities` service and proxies `ElggEntity::can*()` methods.
Adds PHPUnit tests for user capabilities
Creates an easier way to mock entities during testing with `\Elgg\Tests\EntityMocks`
class, which can be instantiated with the instance of PHPUnit test case.

### CODE REVIEW

File: engine/classes/Elgg/UserCapabilities.php [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+	 *
+	 * @param ElggEntity   $entity    Object entity
+	 * @param int          $user_guid The user GUID, optionally (default: logged in user)
+	 * @param ElggMetadata $metadata  The piece of metadata to specifically check or null for any metadata
+	 *
+	 * @return bool
+	 * @see elgg_set_ignore_access()
+	 */
+	public function canEditMetadata(ElggEntity $entity, $user_guid = 0, ElggMetadata $metadata = null) {
+		if (!$entity->guid) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/UserCapabilities.php]: Let's revisit. We should perhaps also check if metadata belongs to this entity in the first place.

File: engine/lib/entities.php [hunk 2]
```diff
@@ -271,43 +271,19 @@ function update_subtype($type, $subtype, $class = '') {
...
 	if (!$container_guid) {
 		$container_guid = elgg_get_page_owner_guid();
 	}
-
 	$container = get_entity($container_guid);
-
-	try {
-		$user = _elgg_services()->entityTable->getUserForPermissionsCheck($user_guid);
-	} catch (UserFetchFailureException $e) {
+	if (!$container) {
```
Comment (hypeJunction) [inline engine/lib/entities.php]: I am not sure how this hasn't xaused problems before.  No page owner would pass empty container to the hook.

File: engine/classes/Elgg/UserCapabilities.php [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+	 *
+	 * @param ElggEntity $entity          Objet entity
+	 * @param int        $user_guid       User guid (default is logged in user)
+	 * @param string     $annotation_name The name of the annotation (default is unspecified)
+	 *
+	 * @return bool
+	 */
+	public function canAnnotate(ElggEntity $entity, $user_guid = 0, $annotation_name = '') {
+		if (!empty($annotation_name) && !is_string($annotation_name)) {
+			throw new \InvalidArgumentException(__METHOD__ . ' expects \$annotation_name to be a string');
```
Comment (hypeJunction) [inline engine/classes/Elgg/UserCapabilities.php]: This is new. I think it makes sense here.

File: engine/classes/Elgg/UserCapabilities.php [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+/**
+ * WARNING: API IN FLUX. DO NOT USE DIRECTLY.
+ * 
+ * Use the elgg_* versions instead.
+ * 
+ * @access private
+ * @since 2.2
+ */
+class UserCapabilities {
```
Comment (mrclay) [inline engine/classes/Elgg/UserCapabilities.php]: Did moving these methods make testing a lot easier?

File: engine/tests/phpunit/Elgg/UserCapabilitiesTest.php [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+				->will($this->returnValueMap([
+							['guid', 36],
+							['owner_guid', 33]
+		]));
+
+		$this->setupServices();
+		$this->assertFalse($user->canComment(33));
+		$this->assertTrue($object->canComment(33));
+		$this->assertFalse($group->canComment(33));
+		$this->assertNull($entity->canComment(33));
```
Comment (mrclay) [inline engine/tests/phpunit/Elgg/UserCapabilitiesTest.php]: Why not false?

File: engine/tests/phpunit/Elgg/UserCapabilitiesTest.php [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+	 * @group UserCapabilities
+	 */
+	public function testCanAnnotateByDefault() {
+		$this->setupEntity(41, 'user', '');
+		$this->setupEntity(42, 'user', '');
+		$entity = $this->setupEntity(43, 'object', 'foo', [
+			'owner_guid' => 41,
+		]);
+		$this->setupServices();
+		$entity->canAnnotate(41); // owner
```
Comment (mrclay) [inline engine/tests/phpunit/Elgg/UserCapabilitiesTest.php]: Missing assertions?

File: engine/classes/Elgg/UserCapabilities.php (hunk 3 above)
Comment (mrclay) [inline engine/classes/Elgg/UserCapabilities.php]: This would still accept an array. Do we really need to accept `null` and `false`?

Comment (mrclay) [inline engine/classes/Elgg/UserCapabilities.php]: Rather be explicit about it:
```php
if ($annotation_name === null || $annotation_name === false) {
    // accepting these for BC
    $annotation_name = '';
} elseif (!is_string($annotation_name)) {
    throw ...
}
```

**Comment by mrclay** [comment]
I like a lot that this centralizes the capabilities logic.

The complexity of the test mocking makes me nervous, but since the capabilities mostly depends on hooks and attributes (not deep queries) I think this is a good addition. Would be nice to remove duplicate test functionality from simpletests. I'll try to identify some of that.

File: engine/tests/phpunit/Elgg/UserCapabilitiesTest.php [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+		if ($type && $entity && $entity->getType() == $type) {
+			return $entity;
+		}
+		return false;
+	}
+
+	/**
+	 * Setup a mock entity
+	 */
+	protected function setupEntity($guid, $type, $subtype, array $attributes = []) {
```
Comment (mrclay) [inline engine/tests/phpunit/Elgg/UserCapabilitiesTest.php]: Can later be moved to an entityMocks service

**Comment by mrclay** [comment]
This should ideally replace:

- [ ] ElggAnnotationTest
- [ ] \ElggCoreEntityTest::testCanAnnotateDefault
- [ ] \ElggCoreEntityTest::testCanAnnotateCallsSpecificThenGenericHook
- [ ] \ElggCoreEntityTest::testCanAnnotateHookParams
- [ ] \ElggCoreEntityTest::testCanAnnotateDoesntCallSpecificThenGenericHookForEmptyString
- [ ] \ElggCoreRegressionBugsTest::test_can_write_to_container

File: engine/classes/Elgg/UserCapabilities.php (hunk 4 above)
Comment (hypeJunction) [inline engine/classes/Elgg/UserCapabilities.php]: It did

File: engine/tests/phpunit/Elgg/UserCapabilitiesTest.php (hunk 7 above)
Comment (hypeJunction) [inline engine/tests/phpunit/Elgg/UserCapabilitiesTest.php]: Yes. This was the biggest challenge, but we now have a foundation to work with

File: engine/tests/phpunit/Elgg/UserCapabilitiesTest.php (hunk 5 above)
Comment (hypeJunction) [inline engine/tests/phpunit/Elgg/UserCapabilitiesTest.php]: I don't know. That's what it is design. The value is then checked by extending classes. I am planning to tackle this wihen I work on commenting permissions.

**Comment by hypeJunction** [comment]
@mrclay 
Simplified mocked by adding an extra class 
Added `canEditAnnotation` and removed redundant annotation tests. 
Updated some uses of `can_write_to_container()` (not really sure what the pages plugin is doing there, so please take a look)
Not touching `\ElggCoreRegressionBugsTest::test_can_write_to_container` as it looks like it's relying on group hooks for write access.

**Comment by hypeJunction** [comment]
Pages plugin is another use case for `container_logic_permissions`.

FYI, I am headed to the airport, and won't be around much this week, so feel free to hijack any open pull requests and do your magic. Bis gleich.

File: engine/tests/phpunit/Elgg/UserCapabilitiesTest.php [hunk 8]
```diff
@@ -0,0 +1,694 @@
...
+		$viewer = $this->mocks->getUser();
+		$this->assertTrue($user->canEdit($user->guid));
+		$this->assertFalse($user->canEdit($viewer->guid));
+	}
+
+	/**
+	 * @group UserCapabilities
+	 * @\expectedException UserFetchFailureException
+	 */
+	public function testThrowsExceptionForCanObjectEditSelf() {
```
Comment (hypeJunction) [inline engine/tests/phpunit/Elgg/UserCapabilitiesTest.php]: @mrclay Could you peep this. I am not sure why the class is not autoloaded.

File: mod/blog/actions/blog/save.php [hunk 9]
```diff
@@ -91,8 +91,8 @@
 		case 'container_guid':
 			// this can't be empty or saving the base entity fails
 			if (!empty($value)) {
-				if (can_write_to_container($user->getGUID(), $value)) {
-					$values[$name] = $value;
```
Comment (hypeJunction) [inline mod/blog/actions/blog/save.php]: Oops

**Comment by mrclay** [comment]
Would this cover #8790?

**Comment by hypeJunction** [comment]
I will need to look after a good night sleep, but my gut says logic check would cover the case. Can try approaching with tests to see what happens.

**Comment by hypeJunction** [comment]
@mrclay I think #8790 is better off with the solution you proposed. Using this would still be misusing container permissions.

**Comment by hypeJunction** [comment]
@mrclay I have removed extra features from here. Let's get this in and then I will make new PRs for the rest of the stuff. It was getting hard to work with multiple commits.

**Comment by mrclay** [comment]
Porting of permissions tests always makes me nervous, so I'm going to split this into 2 PRs, first that adds API and passes old tests, 2nd that ports tests.

**Comment by mrclay** [comment]
Replaced by #9736 and  #9737

**Comment by mrclay** [comment]
...which are both in!

**Comment by hypeJunction** [comment]
Awesome. `container_logic_check` will follow.

File: engine/classes/Elgg/UserCapabilities.php (hunk 1 above)
Comment (hypeJunction) [inline engine/classes/Elgg/UserCapabilities.php]: Will open an issue.

**Pull request closed without merging**